### PR TITLE
Update privateclusters e2e test for SDKv2

### LINKED ISSUE #4056: Convert private clusters e2e tests to SDKv2
/kind bug

**What steps did you take and what happened:**

The last step in removing Azure SDK for Go v1 code from CAPZ is in the private cluster spec of the e2e tests. It's likely that using the default Azure credential factory somehow doesn't work in that more intricate environment.

See the history of #3991 for more details.

### PR DESCRIPTION
**Pull request by mboersma**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Removes the last vestige of Azure SDK v1 code from CAPZ, specifically the `privatecluster` e2e test.

**Which issue(s) this PR fixes**:

Fixes #4056

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by mboersma** [comment]
/retitle [WIP] Update privateclusters e2e test for SDKv2

**Comment by mboersma** [comment]
*All* the -optional test specs failed in the same, apparently unrelated to this change. I'll run them once more.

/test pull-cluster-api-provider-azure-e2e-optional

**Comment by mboersma** [comment]
w00t it passed (ignore the Flatcar failure). I'll squash and promote from draft.

**Comment by mboersma** [comment]
Lots of red on that run, but it did pass the privateclusters e2e spec: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/4268/pull-cluster-api-provider-azure-e2e-optional/1729921750407319552

**Comment by nojnhuh** [comment]
@mboersma For context, what was the heart of the reason that this test was more complicated to update to SDK v2 than other tests?

**Comment by mboersma** [comment]
@nojnhuh the main differences were:
- multiple clients from different services involved
- some generic resource code that can't be replicated in SDKv2 (which I removed)

But overall, the issue was that I previously updated this all at once, in a flurry of test and service refactoring, and it failed in a way I wasn't able to diagnose.

So I tried again, replacing the code bit by bit, and this time it succeeded. It took quite a while because I'm not able to run this particular test locally, so I had to make it a draft PR and then manually run the `-optional` test (and ignore the Flatcar failure).

**Review by nojnhuh — COMMENTED** [review]
/lgtm

File: .golangci.yml [hunk 1]
```diff
@@ -140,10 +140,6 @@ issues:
         - prealloc
     - path: ^test/
       text: exported (.+) should have comment( \(or a comment on this block\))? or be unexported
-    # Need to use the deprecated module "github.com/Azure/azure-sdk-for-go/services" till issue #2670 is addressed.
```
[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply .golangci.yml:143]: 🥳

**Comment by mboersma** [comment]
The `-optional` test failed again, but the privateclusters spec succeeded.

**Comment by CecileRobertMichon** [comment]
Flatcar should be fixed now

/retest

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**